junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TestInstancePreDestroyCallbackTests.java

```diff
@@ -0,0 +1,248 @@
+                "test2",
+                "fooPreDestroyCallbackTestInstance:PerClassLifecyclePreDestroyCallbackWithTwoTestMethods"
+        );
+        // @formatter:on
+	}
+
+	// -------------------------------------------------------------------
+
+	@ExtendWith(FooInstancePostProcessor.class)
+	@ExtendWith(FooInstancePreDestroyCallback.class)
+	static class OuterTestCase extends Named {
+
+		@BeforeEach
+		void beforeOuterMethod() {
+			callSequence.add("beforeOuterMethod");
+		}
+
+		@Test
+		void testOuter() {
+			assertFalse(isDestroyed);
+			callSequence.add("testOuter");
+		}
+
+		@Nested
+		@ExtendWith(BarInstancePostProcessor.class)
+		@ExtendWith(BarInstancePreDestroyCallback.class)
+		class InnerTestCase extends Named {
+
+			@BeforeEach
+			void beforeInnerMethod() {
+				assertFalse(isDestroyed);
+				callSequence.add("beforeInnerMethod");
+			}
+
+			@Test
+			void testInner() {
+				callSequence.add("testInner");
+			}
+		}
+	}
+
+	static class TestCaseWithTestSpecificTestInstancePreDestroyCallback extends Named {
+
+		@BeforeEach
+		void beforeEachMethod() {
+			assertFalse(isDestroyed);
+			callSequence.add("beforeEachMethod");
+		}
+
+		@ExtendWith(FooInstancePostProcessor.class)
+		@ExtendWith(FooInstancePreDestroyCallback.class)
+		@Test
+		void test() {
+			callSequence.add("test");
+		}
+	}
+
+	@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+	@ExtendWith(FooInstancePostProcessor.class)
+	@ExtendWith(FooInstancePreDestroyCallback.class)
+	static class PerClassLifecyclePreDestroyCallbackWithTwoTestMethods extends Named {
+
+		@BeforeEach
+		void beforeEachMethod() {
+			callSequence.add("beforeEachMethod");
+		}
+
+		@Test
+		void test1() {
+			callSequence.add("test1");
+		}
+
+		@Test
+		void test2() {
+			callSequence.add("test2");
+		}
+	}
+
+	static class FooInstancePostProcessor implements TestInstancePostProcessor {
+
+		@Override
+		public void postProcessTestInstance(Object testInstance, ExtensionContext context) {
+			if (testInstance instanceof Named) {
+				((Named) testInstance).setName("foo:" + context.getRequiredTestClass().getSimpleName());
+			}
+			callSequence.add("fooPostProcessTestInstance:" + testInstance.getClass().getSimpleName());
+		}
+	}
+
+	static class BarInstancePostProcessor implements TestInstancePostProcessor {
+
+		@Override
+		public void postProcessTestInstance(Object testInstance, ExtensionContext context) {
+			if (testInstance instanceof Named) {
+				((Named) testInstance).setName("bar:" + context.getRequiredTestClass().getSimpleName());
+			}
+			callSequence.add("barPostProcessTestInstance:" + testInstance.getClass().getSimpleName());
+		}
+	}
+
+	static class FooInstancePreDestroyCallback implements TestInstancePreDestroyCallback {
+
+		@Override
+		public void preDestroyTestInstance(Object testInstance, ExtensionContext context) throws Exception {
+			if (testInstance instanceof Named) {
+				String name;
+				if ("InnerTestCase".equals(testInstance.getClass().getSimpleName())) {
```

Comment:
Maybe you could extract `testInstance.getClass().getSimpleName()` as your using it several times? You could also simplify the if statement with a ternary operator.

Reply:
pending decision on whether to remove PostProcessor 's

Reply:
This one's inside `FooInstancePreDestroyCallback`, so it will stay either way 😉

Reply:
extract out abstract class

Reply:
👍